packages/devtools_app/lib/src/charts/chart_controller.dart

```diff
@@ -0,0 +1,449 @@
+        final lastDT = DateTime.fromMillisecondsSinceEpoch(lastTimestamp);
+        durationLabel = midDT.difference(lastDT);
+      }
+    }
+  }
+
+  /// Used to compute incoming ticks.
+  /// @param refresh true rebuild all labels (probably zoom changed).
+  /// @param timestamp current timestamp received implies building all
+  /// parts of the x-axis labels.
+  void recomputeLabels({int timestamp, bool refresh = false}) {
+    if (refresh) {
+      // Need the correct tickWidth based on current zoom.
+      computeChartArea();
+
+      // All labels need to be recomputed.
+      durationLabel = null;
+      _xAxisLabeledTimestamps.replaceRange(
+        0,
+        _xAxisLabeledTimestamps.length,
+        [null, null, null],
+      );
+
+      final timestampsLength = timestamps.length;
+      final midPt = (visibleTicks / 2).truncate();
+      if (timestampsLength > visibleTicks) {
+        leftLabelTimestamp = timestamps[timestamps.length - visibleTicks + 10];
+        centerLabelTimestamp = timestamps[timestamps.length - midPt];
+        rightLabelTimestamp = timestamps.last;
+      } else if (timestampsLength > midPt) {
+        centerLabelTimestamp = timestamps[timestamps.length - midPt];
+        rightLabelTimestamp = timestamps.last;
+      } else if (timestamps.isNotEmpty) {
+        rightLabelTimestamp = timestamps.first;
+      }
+      return;
+    }
+
+    if (durationLabel == null && rightLabelTimestamp == null) {
+      // No center label so start first label (right-side).
+      rightLabelTimestamp = timestamp;
+    } else if (durationLabel != null && centerLabelTimestamp == null) {
+      // Need a center label are we at the duration we want for the
+      // next tick label?
+      final rightDT = DateTime.fromMillisecondsSinceEpoch(rightLabelTimestamp);
+      final currentDT = DateTime.fromMillisecondsSinceEpoch(timestamp);
+      if (currentDT.difference(rightDT).inSeconds >= durationLabel.inSeconds) {
+        slideLabelsLeft();
+        rightLabelTimestamp = timestamp;
+      }
+    } else if (durationLabel != null && leftLabelTimestamp == null) {
+      // Need a left label are we at the duration we want for the
+      // next tick label?
+      final rightDT = DateTime.fromMillisecondsSinceEpoch(rightLabelTimestamp);
+      final currentDT = DateTime.fromMillisecondsSinceEpoch(timestamp);
+      if (currentDT.difference(rightDT).inSeconds >= durationLabel.inSeconds) {
+        slideLabelsLeft();
+        rightLabelTimestamp = timestamp;
+      }
+    }
+  }
+
+  void slideLabelsLeft() {
+    // Left-side out of visible range. Slide center and right labels to
+    // left-most and center. Need to eventually recompute right-most when
+    // duration gap is met.
+    leftLabelTimestamp = centerLabelTimestamp;
+    centerLabelTimestamp = rightLabelTimestamp;
+    rightLabelTimestamp = null;
+  }
+
+  // xCanvas coord for plotting data.
+  double xCanvasChart;
```

Comment:
one convention from Flutter is to generally use Offset instead of having pairs of variables with names that start with x and y.